File: Data Structure/String/Medium/LET_no890_Find and Replace Pattern.cpp
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> res;
        for (string& w : words) {
            if (w.size() != pattern.size()) continue;
            unordered_map<char, char> w2p, p2w;
            bool valid = true;
            for (int i = 0; i < w.size(); i++) {
                if (!w2p.count(w[i]) and !p2w.count(pattern[i])) {
                    w2p[w[i]] = pattern[i];
                    p2w[pattern[i]] = w[i];
                }
                else {
                    if (w2p.count(w[i]) && w2p[w[i]] != pattern[i] 
                        || p2w.count(pattern[i]) && p2w[pattern[i]] != w[i]) {
                        valid = false;
                        break;
                    }
                }
            }
            if (valid) res.push_back(w);
        }
        
        return res;
    }
};

File: Data Structure/Hash/Medium/LET_no890_Find and Replace Pattern.py
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        res = []
        for w in words:
            if len(w) != len(pattern): continue
            w2p, p2w = dict(), dict()
            valid = True
            for i, j in zip(w, pattern):
                if i not in w2p and j not in p2w:
                    w2p[i] = j
                    p2w[j] = i
                else:
                    if i in w2p and w2p[i] != j or j in p2w and p2w[j] != i: 
                        valid = False
                        break
            if valid: res.append(w)
        
        return res